src/main/java/org/skriptlang/skript/lang/comparator/Comparators.java

```diff
@@ -0,0 +1,241 @@
+	 */
+	private static final List<ComparatorInfo<?, ?>> COMPARATORS = new ArrayList<>();
+
+	/**
+	 * @return An unmodifiable list containing all registered {@link ComparatorInfo}s.
+	 * Please note that this does not include any special Comparators resolved by Skript during runtime.
+	 * This method ONLY returns Comparators explicitly registered during registration.
+	 */
+	public static List<ComparatorInfo<?, ?>> getComparatorInfo() {
+		return Collections.unmodifiableList(COMPARATORS);
+	}
+
+	/**
+	 * A map for quickly accessing comparators that have already been resolved.
+	 * This is useful for skipping complex lookups that may require conversion and inversion.
+	 */
+	private static final Map<Pair<Class<?>, Class<?>>, Comparator<?, ?>> QUICK_ACCESS_COMPARATORS = new HashMap<>();
+
+	/**
+	 * Registers a new Comparator with Skript's collection of Comparators.
+	 * @param firstType The first type for comparison.
+	 * @param secondType The second type for comparison.
+	 * @param comparator A Comparator for comparing objects of 'firstType' and 'secondType'.
+	 */
+	public static <Type1, Type2> void registerComparator(
+		Class<Type1> firstType,
+		Class<Type2> secondType,
+		Comparator<Type1, Type2> comparator
+	) {
+		Skript.checkAcceptRegistrations();
+
+		if (firstType == Object.class && secondType == Object.class)
+			throw new IllegalArgumentException("It is not possible to add a comparator between objects");
+
+		COMPARATORS.add(new ComparatorInfo<>(firstType, secondType, comparator));
+	}
+
+	/**
+	 * Compares two objects to see if a Relation exists between them.
+	 * @param first The first object for comparison.
+	 * @param second The second object for comparison.
+	 * @return The Relation between the two provided objects.
+	 */
+	@SuppressWarnings("unchecked")
+	public static <Type1, Type2> Relation compare(@Nullable Type1 first, @Nullable Type2 second) {
+		if (first == null || second == null)
+			return Relation.NOT_EQUAL;
+
+		Comparator<Type1, Type2> comparator = getComparator((Class<Type1>) first.getClass(), (Class<Type2>) second.getClass());
+		if (comparator == null)
+			return Relation.NOT_EQUAL;
+
+		return comparator.compare(first, second);
+	}
+
+	/**
+	 * A method for obtaining a comparator that can compare two objects of 'firstType' and 'secondType'.
+	 * Please note that comparators may convert objects if necessary for comparisons.
+	 * @param firstType The first type for comparison.
+	 * @param secondType The second type for comparison.
+	 * @return A comparator capable of determine the {@link Relation} between two objects of 'firstType' and 'secondType'.
+	 * Will be null if no comparator capable of comparing two objects of 'firstType' and 'secondType' was found.
+	 */
+	@Nullable
+	@SuppressWarnings("unchecked")
+	public static <Type1, Type2> Comparator<Type1, Type2> getComparator(Class<Type1> firstType, Class<Type2> secondType) {
+		Pair<Class<?>, Class<?>> typePair = new Pair<>(firstType, secondType);
```

Comment:
You can speed this up and make it simpler.
Back up in ComparatorInfo you should add a hashCode based on (only) the two classes.
You can then search by the hash to find an exact match a lot quicker.

Reply:
Clearly, a Pair does the same thing eventually, but it's requiring the creation of a new thing when you can use an integer instead.